src/keydra/providers/github.py

```diff
@@ -0,0 +1,238 @@
+                dest['key'],
+                ' ({})'.format(secret['key']) if 'key' in secret else '',
+                dest['source'],
+                secret['provider'],
+                config['repository'],
+            )
+        )
+
+        return dest
+
+    def rotate(self, secret):
+        raise RotationException('Github provider does not support rotation')
+
+    def _distribute(self, secret, destination):
+        try:
+            if destination['scope'] == 'repository':
+                return self._distribute_repository_secret(secret, destination)
+
+            else:
+                raise NotImplementedError(
+                    'Github for scope {}'.format(destination['scope'])
+                )
+
+        except Exception as e:
+            raise DistributionException(e)
+
+    @exponential_backoff_retry(3)
+    def distribute(self, secret, destination):
+        return self._distribute(secret, destination)
+
+    @classmethod
+    def _validate_repository_spec(cls, spec):
+        if 'repository' not in spec['config']:
+            return False, 'Attribute "config.repository" not present'
+
+        return True, ''
+
+    @classmethod
+    def validate_spec(cls, spec):
+        valid, msg = BaseProvider.validate_spec(spec)
+
+        if not valid:
+            return valid, msg
+
+        if 'scope' not in spec:
+            return False, 'Attribute "scope" not present in configuration'
+
+        if 'config' not in spec:
+            return False, 'Attribute "config" not present in configuration'
+
+        if spec['scope'] != 'repository':
+            return False, 'Unsupported scope'
+        else:
+            return Client._validate_repository_spec(spec)
+
+        return True, 'It is valid!'  # pragma: no cover
+
+    @classmethod
+    def pre_process_spec(self, spec, context={}):
+        specs = []
+
+        target = copy.deepcopy(spec)
+
+        current_env_name, current_env = list(
+            context.get('environment', {}).items()
+        )[0]
+
+        if 'ENV' in target['key']:
+            target['key'] = target['key'].format(
+                **{'ENV': current_env_name.upper()}
+            )
+
+        if 'ENV' in target.get('config', {}).get(
+            'environment', ''
+        ):
+            target['config']['environment'] = target[
+                'config'
+            ]['environment'].format(
+                **{'ENV': current_env_name.upper()}
+            )
+
+        if spec['scope'] in ['deployment', 'repository'] and any(
+            [
+                isinstance(spec['config']['repository'], list),
+                isinstance(spec['config']['repository'], tuple)
+            ]
+        ):
+            for repo in spec['config']['repository']:
+                new_target = copy.deepcopy(target)
+                new_target['config']['repository'] = repo
+
+                specs.append(new_target)
+        else:
+            specs.append(target)
+
+        return specs
```

Comment:
To be honest I never fully understood this ENV black magic as a user writing Keydra configs. I'm afraid looking at the implementation doesn't make it immediately clearer 😄 But it looks like a copy of Bitbucket provider, so at least it should provide like-for-like functionality..

Reply:
Agree, I've struggled too. Will look to change this in future to make it easier to understand.